From: Pierre Ynard <linkfanel@yahoo.fr>
Date: Wed, 4 Jan 2017 01:45:35 +0000 (+0100)
Subject: rdnssd: properly handle search list entries in merge hook
X-Git-Url: http://git.remlab.net/gitweb/?p=ndisc6.git;a=commitdiff_plain;h=d60853a5319bac0c3ec9a082bcaf850a5ab8d1d5

rdnssd: properly handle search list entries in merge hook

Basically, drop DNSSL entries because the hook is too basic to handle
them correctly. Use something more sophisticated like resolvconf if you
want this functionality. This fixes the following issues:

 - inserting less IPv6 nameservers than calculated
 - littering /etc/resolv.conf with stray search lines every time DNSSL
   entries change
 - clobbering existing (DHCPv4) search list entries
 - overkill use of /usr/bin/awk, outside of PATH and reliant on /usr
   availability

Signed-off-by: Rémi Denis-Courmont <remi@remlab.net>
---

diff --git a/rdnssd/merge-hook.in b/rdnssd/merge-hook.in
index 2d202e8..383a57c 100644
--- a/rdnssd/merge-hook.in
+++ b/rdnssd/merge-hook.in
@@ -3,7 +3,7 @@
 # resolv.conf merge hook for rdnssd
 
 # *************************************************************************
-# *  Copyright © 2007-2009 Pierre Ynard.                                  *
+# *  Copyright © 2007-2009, 2017 Pierre Ynard.                            *
 # *  This program is free software: you can redistribute and/or modify    *
 # *  it under the terms of the GNU General Public License as published by *
 # *  the Free Software Foundation, versions 2 or 3 of the license.        *
@@ -51,13 +51,18 @@ if [ $limit -lt $room ]; then
 	limit=$room
 fi
 
-# Merge and write the result
+# Merge and write the result. Let rdnssd assume ownership of all IPv6
+# nameservers, and remove extraneous IPv6 entries as expired. However
+# DHCPv4 most often sets up search list entries, and rdnssd cannot
+# clobber these lest it causes counterintuitive breakage. There is no
+# easy way to properly merge and manage DNSSL entries here, so just drop
+# them.
 
 {
 	sed -e "/$RE_NSV4OR6/d" < $resolvconf
-        [ $limit -gt 0 ] && sed -e "${limit}q" < $myresolvconf
+	grep -m $limit "$RE_NSV4OR6" < $myresolvconf || [ $? -le 1 ]
 	sed -ne "/$RE_NSV4/p" < $resolvconf
-} | awk '!a[$0]++' > $resolvconf.tmp
+} > $resolvconf.tmp
 
 mv -f $resolvconf.tmp $resolvconf
 
